redis.go

```diff
@@ -1,46 +1,96 @@
+// Package redis implements a redis client for k6.
 package redis
 
 import (
-	"context"
-	"time"
+	"errors"
+	"strconv"
 
+	"github.com/dop251/goja"
 	"github.com/go-redis/redis/v8"
-
 	"go.k6.io/k6/js/common"
 	"go.k6.io/k6/js/modules"
 )
 
 // Register the extension on module initialization, available to
 // import from JS as "k6/x/redis".
 func init() {
-	modules.Register("k6/x/redis", new(Redis))
+	modules.Register("k6/x/redis", New())
 }
 
-// Redis is the k6 extension for a Redis client.
-type Redis struct{}
+type (
+	// RootModule is the global module instance that will create Client
+	// instances for each VU.
+	RootModule struct{}
+
+	// ModuleInstance represents an instance of the JS module.
+	ModuleInstance struct {
+		vu modules.VU
+
+		*Client
+	}
+)
+
+// Ensure the interfaces are implemented correctly
+var (
+	_ modules.Instance = &ModuleInstance{}
+	_ modules.Module   = &RootModule{}
+)
 
-// Client is the Redis client wrapper.
-type Client struct {
-	client *redis.Client
+// New returns a pointer to a new RootModule instance
+func New() *RootModule {
+	return &RootModule{}
 }
 
-// XClient represents the Client constructor (i.e. `new redis.Client()`) and
-// returns a new Redis client object.
-func (r *Redis) XClient(ctxPtr *context.Context, opts *redis.Options) interface{} {
-	rt := common.GetRuntime(*ctxPtr)
-	return common.Bind(rt, &Client{client: redis.NewClient(opts)}, ctxPtr)
+// NewModuleInstance implements the modules.Module interface and returns
+// a new instance for each VU.
+func (*RootModule) NewModuleInstance(vu modules.VU) modules.Instance {
+	return &ModuleInstance{vu: vu, Client: &Client{vu: vu}}
 }
 
-// Set the given key with the given value and expiration time.
-func (c *Client) Set(key, value string, exp time.Duration) {
-	c.client.Set(c.client.Context(), key, value, exp)
+// Exports implements the modules.Instance interface and returns
+// the exports of the JS module.
+func (mi *ModuleInstance) Exports() modules.Exports {
+	return modules.Exports{Named: map[string]interface{}{
+		"Client": mi.NewClient,
+	}}
 }
 
-// Get returns the value for the given key.
-func (c *Client) Get(key string) (string, error) {
-	res, err := c.client.Get(c.client.Context(), key).Result()
-	if err != nil {
-		return "", err
+// NewClient is the JS constructor for the redis Client.
+func (mi *ModuleInstance) NewClient(call goja.ConstructorCall) *goja.Object {
+	rt := mi.vu.Runtime()
+
+	addrArg := call.Argument(0).String()
```

Comment:
Since you also asked this in the PR, so you need some struct
```go
type Configuration struct {
   Foo string
   Bar string
}
```
and use something like:
```go
var cfg Configuration
if err := rt.ExportTo(call.Argument(0), &cfg); err != nil {
   common.Throw(rt, fmt.Errorf("Client constructor expects Configuration as it's argument: %w", err))
}
```

Reply:
AWESOME! Vielen Dank 👏🏻